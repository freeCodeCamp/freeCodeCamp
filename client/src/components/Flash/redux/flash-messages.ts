export enum FlashMessages {
  AccountDeleted = 'flash.account-deleted',
  AddNameSuccess = 'flash.add-name',
  AlreadyClaimed = 'flash.already-claimed',
  CertClaimSuccess = 'flash.cert-claim-success',
  CertificateMissing = 'flash.certificate-missing',
  CertsPrivate = 'flash.certs-private',
  ChallengeSaveTooBig = 'flash.challenge-save-too-big',
  ChallengeSubmitTooBig = 'flash.challenge-submit-too-big',
  CodeSaved = 'flash.code-saved',
  CodeSaveError = 'flash.code-save-error',
  CodeSaveLess = 'flash.code-save-less',
  CompleteProjectFirst = 'flash.complete-project-first',
  CourseUrlCopied = 'flash.course-url-copied',
  CourseUrlCopyError = 'flash.course-url-copy-error',
  DeleteTokenErr = 'flash.delete-token-err',
  EmailValid = 'flash.email-valid',
  GenerateExamError = 'flash.generate-exam-error',
  HonestFirst = 'flash.honest-first',
  IncompleteSteps = 'flash.incomplete-steps',
  LocalCodeSaved = 'flash.local-code-saved',
  LocalCodeSaveError = 'flash.local-code-save-error',
  MsTranscriptErr1 = 'flash.ms.transcript.link-err-1',
  MsTranscriptErr2 = 'flash.ms.transcript.link-err-2',
  MsTranscriptErr3 = 'flash.ms.transcript.link-err-3',
  MsTranscriptErr4 = 'flash.ms.transcript.link-err-4',
  MsTranscriptErr5 = 'flash.ms.transcript.link-err-5',
  MsTranscriptErr6 = 'flash.ms.transcript.link-err-6',
  MsTranscriptLinked = 'flash.ms.transcript.linked',
  MsTranscriptUnlinked = 'flash.ms.transcript.unlinked',
  MsTranscriptUnlinkErr = 'flash.ms.transcript.unlink-err',
  MsProfileErr = 'flash.ms.profile.err',
  MsTrophyErr1 = 'flash.ms.trophy.err-1',
  MsTrophyErr2 = 'flash.ms.trophy.err-2',
  MsTrophyErr3 = 'flash.ms.trophy.err-3',
  MsTrophyErr4 = 'flash.ms.trophy.err-4',
  MsTrophyErr5 = 'flash.ms.trophy.err-5',
  MsTrophyErr6 = 'flash.ms.trophy.err-6',
  MsTrophyVerified = 'flash.ms.trophy.verified',
  NameNeeded = 'flash.name-needed',
  None = '',
  NotEligible = 'flash.not-eligible',
  NotHonest = 'flash.not-honest',
  GenericError = 'flash.generic-error',
  ProfilePrivate = 'flash.profile-private',
  ProgressReset = 'flash.progress-reset',
  ProvideUsername = 'flash.provide-username',
  ReallyWeird = 'flash.really-weird',
  ReportSent = 'flash.report-sent',
  SigninSuccess = 'flash.signin-success',
  StartProjectErr = 'flash.start-project-err',
  SurveyErr1 = 'flash.survey.err-1',
  SurveyErr2 = 'flash.survey.err-2',
  SurveyErr3 = 'flash.survey.err-3',
  SurveySuccess = 'flash.survey.success',
  TimelinePrivate = 'flash.timeline-private',
  TokenDeleted = 'flash.token-deleted',
  UpdatedAboutMe = 'flash.updated-about-me',
  UpdatedSocials = 'flash.updated-socials',
  UpdatedSound = 'flash.updated-sound',
  UpdatedThemes = 'flash.updated-themes',
  UpdatedKeyboardShortCuts = 'flash.keyboard-shortcut-updated',
  UpdatedQunicyEmailSubscription = 'flash.subscribe-to-quincy-updated',
  UpdatedPortfolio = 'flash.portfolio-item-updated',
  UpdatedPrivacy = 'flash.privacy-updated',
  UsernameNotFound = 'flash.username-not-found',
  UsernameTaken = 'flash.username-taken',
  UsernameUpdated = 'flash.username-updated',
  UsernameUsed = 'flash.username-used',
  UserNotCertified = 'flash.user-not-certified',
  UserTokenCopied = 'flash.user-token-copied',
  UserTokenCopyError = 'flash.user-token-copy-error',
  UserTokenGenerated = 'flash.user-token-generated',
  UserTokenGenerateError = 'flash.user-token-generate-error',
  WrongName = 'flash.wrong-name',
  WrongUpdating = 'flash.wrong-updating',
  WentWrong = 'flash.went-wrong'
}
