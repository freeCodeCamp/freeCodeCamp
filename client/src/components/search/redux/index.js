import { createAction, handleActions } from 'redux-actions';

import { createTypes } from '../../../utils/create-types';

export const ns = 'search';

const initialState = {
  query: '',
  indexName: 'news',
  isSearchDropdownEnabled: true,
  isSearchBarFocused: false
};

const types = createTypes(
  ['toggleSearchDropdown', 'toggleSearchFocused', 'updateSearchQuery'],
  ns
);

export const toggleSearchDropdown = createAction(types.toggleSearchDropdown);
export const toggleSearchFocused = createAction(types.toggleSearchFocused);
export const updateSearchQuery = createAction(types.updateSearchQuery);

export const isSearchDropdownEnabledSelector = state =>
  state[ns].isSearchDropdownEnabled;
export const isSearchBarFocusedSelector = state => state[ns].isSearchBarFocused;
export const searchQuerySelector = state => state[ns].query;

export const reducer = handleActions(
  {
    [types.toggleSearchDropdown]: (state, { payload }) => ({
      ...state,
      isSearchDropdownEnabled:
        typeof payload === 'boolean' ? payload : !state.isSearchDropdownEnabled
    }),
    [types.toggleSearchFocused]: (state, { payload }) => {
      const focussed = state.isSearchBarFocused;
      return focussed === payload
        ? state
        : { ...state, isSearchBarFocused: payload };
    },
    [types.updateSearchQuery]: (state, { payload }) => ({
      ...state,
      query: payload
    })
  },
  initialState
);
