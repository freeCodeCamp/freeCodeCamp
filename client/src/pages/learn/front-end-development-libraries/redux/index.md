---
title: Introduction to the Redux Challenges
block: redux
superBlock: front-end-development-libraries
---

## Introduction to the Redux Challenges

[Redux](https://redux.js.org/) is a predictable state container for JavaScript apps. It helps you write applications that behave consistently, run in different environments (client, server, and native), and are easy to test. While you can use Redux with any view library, it's introduced here before being combined with React.
