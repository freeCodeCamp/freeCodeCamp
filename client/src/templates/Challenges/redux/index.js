import { isEmpty } from 'lodash-es';
import { handleActions } from 'redux-actions';

import { getLines } from '../../../../../shared/utils/get-lines';
import { mergeChallengeFiles } from '../classic/saved-challenges';
import { getTargetEditor } from '../utils/get-target-editor';
import { actionTypes, ns } from './action-types';
import codeStorageEpic from './code-storage-epic';
import completionEpic from './completion-epic';
import createQuestionEpic from './create-question-epic';
import { createCurrentChallengeSaga } from './current-challenge-saga';
import { createExecuteChallengeSaga } from './execute-challenge-saga';

export { ns };

const initialState = {
  canFocusEditor: true,
  attempts: 0,
  visibleEditors: {},
  challengeFiles: [],
  challengeMeta: {
    superBlock: '',
    block: '',
    blockHashSlug: '/',
    id: '',
    isLastChallengeInBlock: false,
    nextChallengePath: '/',
    prevChallengePath: '/',
    challengeType: -1
  },
  challengeTests: [],
  consoleOut: [],
  userCompletedExam: null,
  hasCompletedBlock: false,
  isBuildEnabled: true,
  isExecuting: false,
  isResetting: false,
  logsOut: [],
  modal: {
    completion: false,
    help: false,
    video: false,
    reset: false,
    exitExam: false,
    finishExam: false,
    exitQuiz: false,
    finishQuiz: false,
    examResults: false,
    survey: false,
    projectPreview: false,
    shortcuts: false
  },
  portalWindow: null,
  showPreviewPortal: false,
  showPreviewPane: true,
  projectFormValues: {},
  successMessage: 'Happy Coding!',
  isAdvancing: false,
  chapterSlug: '',
  isSubmitting: false
};

export const epics = [completionEpic, createQuestionEpic, codeStorageEpic];

export const sagas = [
  ...createExecuteChallengeSaga(actionTypes),
  ...createCurrentChallengeSaga(actionTypes)
];

export const reducer = handleActions(
  {
    [actionTypes.submitChallenge]: state => ({
      ...state,
      isSubmitting: true
    }),
    [actionTypes.submitChallengeComplete]: state => ({
      ...state,
      isSubmitting: false
    }),
    [actionTypes.submitChallengeError]: state => ({
      ...state,
      isSubmitting: false
    }),
    [actionTypes.createFiles]: (state, { payload }) => ({
      ...state,
      challengeFiles: payload
    }),
    [actionTypes.updateFile]: (
      state,
      { payload: { fileKey, editorValue, editableRegionBoundaries } }
    ) => {
      const updates = {};
      // if a given part of the payload is null, we leave that part of the state
      // unchanged
      if (editableRegionBoundaries !== null)
        updates.editableRegionBoundaries = editableRegionBoundaries;
      if (editorValue !== null) updates.contents = editorValue;
      if (editableRegionBoundaries !== null && editorValue !== null)
        updates.editableContents = getLines(
          editorValue,
          editableRegionBoundaries
        );
      return {
        ...state,
        challengeFiles: state.challengeFiles.map(challengeFile =>
          challengeFile.fileKey === fileKey
            ? { ...challengeFile, ...updates }
            : { ...challengeFile }
        ),
        isBuildEnabled: true
      };
    },
    [actionTypes.storedCodeFound]: (state, { payload }) => ({
      ...state,
      challengeFiles: state.challengeFiles.length
        ? mergeChallengeFiles(state.challengeFiles, payload)
        : payload
    }),
    [actionTypes.initTests]: (state, { payload }) => ({
      ...state,
      challengeTests: payload
    }),
    [actionTypes.initHooks]: (state, { payload }) => ({
      ...state,
      challengeHooks: payload
    }),
    [actionTypes.updateTests]: (state, { payload }) => ({
      ...state,
      challengeTests: payload
    }),
    [actionTypes.initConsole]: (state, { payload }) => ({
      ...state,
      consoleOut: payload ? [payload] : []
    }),
    [actionTypes.updateConsole]: (state, { payload }) => ({
      ...state,
      consoleOut: state.consoleOut.concat(payload)
    }),
    [actionTypes.initLogs]: state => ({
      ...state,
      logsOut: []
    }),
    [actionTypes.updateLogs]: (state, { payload }) => ({
      ...state,
      logsOut: state.logsOut.concat(payload)
    }),
    [actionTypes.logsToConsole]: (state, { payload }) => ({
      ...state,
      consoleOut: isEmpty(state.logsOut)
        ? state.consoleOut
        : state.consoleOut.concat(payload, state.logsOut)
    }),
    [actionTypes.initVisibleEditors]: state => ({
      ...state,
      visibleEditors: { [getTargetEditor(state.challengeFiles)]: true }
    }),
    [actionTypes.updateChallengeMeta]: (state, { payload }) => ({
      ...state,
      challengeMeta: { ...payload }
    }),
    [actionTypes.resetChallenge]: state => {
      const challengeFilesReset = state.challengeFiles.map(challengeFile => ({
        ...challengeFile,
        contents: challengeFile.seed.slice(),
        editableContents: getLines(
          challengeFile.seed,
          challengeFile.seedEditableRegionBoundaries
        ),
        editableRegionBoundaries:
          challengeFile.seedEditableRegionBoundaries.slice()
      }));
      return {
        ...state,
        currentTab: 2,
        challengeFiles: challengeFilesReset,
        challengeTests: state.challengeTests.map(({ text, testString }) => ({
          text,
          testString
        })),
        consoleOut: [],
        isResetting: true,
        attempts: 0
      };
    },
    [actionTypes.resetAttempts]: state => ({
      ...state,
      attempts: 0
    }),
    [actionTypes.stopResetting]: state => ({
      ...state,
      isResetting: false
    }),
    [actionTypes.updateSolutionFormValues]: (state, { payload }) => ({
      ...state,
      projectFormValues: payload
    }),
    [actionTypes.disableBuildOnError]: state => ({
      ...state,
      isBuildEnabled: false
    }),
    [actionTypes.setShowPreviewPortal]: (state, { payload }) => ({
      ...state,
      showPreviewPortal: payload
    }),
    [actionTypes.setShowPreviewPane]: (state, { payload }) => ({
      ...state,
      showPreviewPane: payload
    }),
    [actionTypes.storePortalWindow]: (state, { payload }) => ({
      ...state,
      portalWindow: payload
    }),
    [actionTypes.removePortalWindow]: state => ({
      ...state,
      portalWindow: null
    }),
    [actionTypes.updateSuccessMessage]: (state, { payload }) => ({
      ...state,
      successMessage: payload
    }),
    [actionTypes.setIsAdvancing]: (state, { payload }) => ({
      ...state,
      isAdvancing: payload
    }),
    [actionTypes.setChapterSlug]: (state, { payload }) => ({
      ...state,
      chapterSlug: payload
    }),
    [actionTypes.setUserCompletedExam]: (state, { payload }) => ({
      ...state,
      userCompletedExam: payload
    }),
    [actionTypes.closeModal]: (state, { payload }) => ({
      ...state,
      modal: {
        ...state.modal,
        [payload]: false
      }
    }),
    [actionTypes.openModal]: (state, { payload }) => ({
      ...state,
      modal: {
        ...state.modal,
        [payload]: true
      }
    }),
    [actionTypes.executeChallenge]: state => ({
      ...state,
      currentTab: 3,
      attempts: state.attempts + 1,
      isExecuting: true
    }),
    [actionTypes.executeChallengeComplete]: state => ({
      ...state,
      isExecuting: false
    }),
    [actionTypes.setEditorFocusability]: (state, { payload }) => ({
      ...state,
      canFocusEditor: payload
    }),
    [actionTypes.toggleVisibleEditor]: (state, { payload }) => {
      return {
        ...state,
        visibleEditors: {
          ...state.visibleEditors,
          [payload]: !state.visibleEditors[payload]
        }
      };
    },
    [actionTypes.createQuestion]: (state, { payload }) => ({
      ...state,
      description: payload
    })
  },
  initialState
);
