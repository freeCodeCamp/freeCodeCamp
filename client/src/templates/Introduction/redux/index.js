import { createAction, handleActions } from 'redux-actions';

import { createTypes } from '../../../utils/create-types';

export const ns = 'curriculumMap';

const initialState = {
  expandedState: {
    block: {}
  }
};

const types = createTypes(['resetExpansion', 'toggleBlock'], ns);

export const resetExpansion = createAction(types.resetExpansion);
export const toggleBlock = createAction(types.toggleBlock);

export const makeExpandedBlockSelector = block => state =>
  !!state[ns].expandedState.block[block];

export const reducer = handleActions(
  {
    [types.resetExpansion]: state => ({
      ...state,
      expandedState: {
        block: {}
      }
    }),
    [types.toggleBlock]: (state, { payload }) => ({
      ...state,
      expandedState: {
        ...state.expandedState,
        block: {
          ...state.expandedState.block,
          [payload]: !state.expandedState.block[payload]
        }
      }
    })
  },
  initialState
);
