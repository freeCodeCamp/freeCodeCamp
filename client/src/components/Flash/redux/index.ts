import { nanoid } from 'nanoid';

import {
  FlashState,
  State,
  FlashApp,
  FlashMessageArg,
  LocalStorageThemes
} from '../../../redux/types';
import { playTone } from '../../../utils/tone';
import { FlashMessages } from './flash-messages';

export const flashMessageSelector = (state: State): FlashState['message'] =>
  state[FlashApp].message;

// ACTION DEFINITIONS

enum FlashActionTypes {
  CreateFlashMessage = 'createFlashMessage',
  RemoveFlashMessage = 'removeFlashMessage'
}

const initialState = {
  message: {
    id: '',
    type: '',
    message: FlashMessages.None
  }
};

export const createFlashMessage = (
  flash: FlashMessageArg
): ReducerPayload<FlashActionTypes.CreateFlashMessage> => {
  // Nightmode theme has special tones
  if (flash.variables?.theme) {
    void playTone(flash.variables.theme as LocalStorageThemes);
  } else if (flash.message !== FlashMessages.None) {
    void playTone(flash.message);
  }
  return {
    type: FlashActionTypes.CreateFlashMessage,
    payload: { ...flash, id: nanoid() }
  };
};

export const removeFlashMessage =
  (): ReducerPayload<FlashActionTypes.RemoveFlashMessage> => ({
    type: FlashActionTypes.RemoveFlashMessage
  });

// REDUCER
type ReducerBase<T> = { type: T };
type ReducerPayload<T extends FlashActionTypes> =
  T extends FlashActionTypes.CreateFlashMessage
    ? ReducerBase<T> & {
        payload: FlashState['message'];
      }
    : ReducerBase<T>;

// Does reducer return FlashState or AppState (whole app)?
export const reducer = (
  state: FlashState = initialState,
  action: ReducerPayload<FlashActionTypes>
): FlashState => {
  switch (action.type) {
    case FlashActionTypes.CreateFlashMessage:
      return { ...state, message: action.payload };
    case FlashActionTypes.RemoveFlashMessage:
      return { ...state, message: initialState.message };
    default:
      return state;
  }
};
