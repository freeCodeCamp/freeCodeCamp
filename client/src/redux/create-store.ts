import { createStore as reduxCreateStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from '@redux-devtools/extension';
import { createEpicMiddleware } from 'redux-observable';
import createSagaMiddleware from 'redux-saga';

import envData from '../../config/env.json';
import { isBrowser } from '../../utils';
import rootEpic from './root-epic';
import rootReducer from './root-reducer';
import rootSaga from './root-saga';

declare const module: {
  hot?: {
    accept: (path: string, callback: () => void) => void;
  };
};

const { environment } = envData;

const clientSide = isBrowser();

const sagaMiddleware = createSagaMiddleware({
  context: {
    document: clientSide ? document : {}
  }
});
const epicMiddleware = createEpicMiddleware({
  dependencies: {
    window: clientSide ? window : {},
    location: clientSide ? window.location : {},
    document: clientSide ? document : {}
  }
});

const composeEnhancers = composeWithDevTools({
  // options like actionSanitizer, stateSanitizer
});

export const createStore = (preloadedState = {}) => {
  let store;
  if (environment === 'production') {
    store = reduxCreateStore(
      rootReducer,
      preloadedState,
      applyMiddleware(sagaMiddleware, epicMiddleware)
    );
  } else {
    store = reduxCreateStore(
      rootReducer,
      preloadedState,
      composeEnhancers(applyMiddleware(sagaMiddleware, epicMiddleware))
    );
  }
  sagaMiddleware.run(rootSaga);
  // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
  epicMiddleware.run(rootEpic);

  if (module.hot) {
    // Enable Webpack hot module replacement for reducers
    module.hot.accept('./root-reducer', () => {
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-require-imports
      const nextRootReducer = require('./root-reducer');
      // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
      store.replaceReducer(nextRootReducer);
    });
  }
  return store;
};
