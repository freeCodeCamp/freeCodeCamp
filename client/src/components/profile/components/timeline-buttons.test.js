import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';

import { createStore } from '../../../redux/create-store';
import completedChallenges from '../../../__mocks__/completed-challenges.json';
import Timeline from './time-line';

Date.prototype.toLocaleString = jest.fn(() => 'Dec 29, 2022');
Date.prototype.toISOString = jest.fn(() => '2016-09-28T20:31:56.730Z');

jest.mock('../../../analytics');

jest.mock('gatsby', () => {
  const edges = require('../../../__mocks__/edges.json');
  const React = require('react');
  const gatsby = jest.requireActual('gatsby');
  return {
    ...gatsby,
    useStaticQuery: jest.fn().mockReturnValue({
      allChallengeNode: {
        edges: edges
      }
    }),
    graphql: jest.fn(),
    Link: jest
      .fn()
      .mockImplementation(
        ({
          activeClassName,
          activeStyle,
          getProps,
          innerRef,
          partiallyActive,
          ref,
          replace,
          to,
          ...rest
        }) =>
          React.createElement('a', {
            ...rest,
            href: to,
            gatsby: 'true'
          })
      )
  };
});

it('should check certification page consistency', () => {
  const tree = renderer
    .create(
      <Provider store={createStore()}>
        <Timeline
          completedMap={completedChallenges}
          username='CeritifedUser'
          onPress={() => {}}
        />
      </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
