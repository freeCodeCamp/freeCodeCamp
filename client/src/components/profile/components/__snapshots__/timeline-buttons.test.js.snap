// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should check certification page consistency 1`] = `
<div
  className="mx-[-15px] "
>
  <div
    className=" min-h-[1px] px-[15px] w-full md:w-2/3    md:ml-[16.6%]   "
  >
    <section
      className="card"
    >
      <h2>
        profile.timeline
      </h2>
      <div
        className="h-[20px]"
      />
      <table
        className="table-auto w-full max-w-full border-collapse text-start text-foreground-tertiary [&_th]:font-normal [&_td]:p-1 [&_th]:p-1 [&>tbody>tr:nth-of-type(odd)]:bg-background-tertiary"
      >
        <thead>
          <tr>
            <th>
              profile.challenge
            </th>
            <th>
              settings.labels.solution
            </th>
            <th
              className="text-center"
            >
              profile.completed
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            className="timeline-row"
          >
            <td>
              <a
                gatsby="true"
                href="/learn/2022/responsive-web-design/learn-accessibility-by-building-a-quiz/step-41"
              >
                Step 41
              </a>
            </td>
            <td />
            <td
              className="text-center"
            >
              <time
                dateTime="2016-09-28T20:31:56.730Z"
              >
                Dec 29, 2022
              </time>
            </td>
          </tr>
          <tr
            className="timeline-row"
          >
            <td>
              <a
                gatsby="true"
                href="/learn/relational-database/build-a-number-guessing-game-project/build-a-number-guessing-game"
              >
                Build a Number Guessing Game
              </a>
            </td>
            <td />
            <td
              className="text-center"
            >
              <time
                dateTime="2016-09-28T20:31:56.730Z"
              >
                Dec 29, 2022
              </time>
            </td>
          </tr>
          <tr
            className="timeline-row"
          >
            <td>
              <a
                gatsby="true"
                href="/learn/relational-database/build-a-periodic-table-database-project/build-a-periodic-table-database"
              >
                Build a Periodic Table Database
              </a>
            </td>
            <td />
            <td
              className="text-center"
            >
              <time
                dateTime="2016-09-28T20:31:56.730Z"
              >
                Dec 29, 2022
              </time>
            </td>
          </tr>
          <tr
            className="timeline-row"
          >
            <td>
              <a
                gatsby="true"
                href="/learn/relational-database/learn-git-by-building-an-sql-reference-object/build-an-sql-reference-object"
              >
                Build an SQL Reference Object
              </a>
            </td>
            <td />
            <td
              className="text-center"
            >
              <time
                dateTime="2016-09-28T20:31:56.730Z"
              >
                Dec 29, 2022
              </time>
            </td>
          </tr>
          <tr
            className="timeline-row"
          >
            <td>
              <a
                gatsby="true"
                href="/learn/relational-database/learn-nano-by-building-a-castle/build-a-castle"
              >
                Build a Castle
              </a>
            </td>
            <td />
            <td
              className="text-center"
            >
              <time
                dateTime="2016-09-28T20:31:56.730Z"
              >
                Dec 29, 2022
              </time>
            </td>
          </tr>
          <tr
            className="timeline-row"
          >
            <td>
              <a
                gatsby="true"
                href="/learn/relational-database/build-a-salon-appointment-scheduler-project/build-a-salon-appointment-scheduler"
              >
                Build a Salon Appointment Scheduler
              </a>
            </td>
            <td />
            <td
              className="text-center"
            >
              <time
                dateTime="2016-09-28T20:31:56.730Z"
              >
                Dec 29, 2022
              </time>
            </td>
          </tr>
          <tr
            className="timeline-row"
          >
            <td>
              <a
                gatsby="true"
                href="/learn/relational-database/learn-bash-and-sql-by-building-a-bike-rental-shop/build-a-bike-rental-shop"
              >
                Build a Bike Rental Shop
              </a>
            </td>
            <td />
            <td
              className="text-center"
            >
              <time
                dateTime="2016-09-28T20:31:56.730Z"
              >
                Dec 29, 2022
              </time>
            </td>
          </tr>
          <tr
            className="timeline-row"
          >
            <td>
              <a
                gatsby="true"
                href="/learn/relational-database/learn-advanced-bash-by-building-a-kitty-ipsum-translator/build-a-kitty-ipsum-translator"
              >
                Build a Kitty Ipsum Translator
              </a>
            </td>
            <td />
            <td
              className="text-center"
            >
              <time
                dateTime="2016-09-28T20:31:56.730Z"
              >
                Dec 29, 2022
              </time>
            </td>
          </tr>
          <tr
            className="timeline-row"
          >
            <td>
              <a
                gatsby="true"
                href="/learn/relational-database/build-a-world-cup-database-project/build-a-world-cup-database"
              >
                Build a World Cup Database
              </a>
            </td>
            <td />
            <td
              className="text-center"
            >
              <time
                dateTime="2016-09-28T20:31:56.730Z"
              >
                Dec 29, 2022
              </time>
            </td>
          </tr>
          <tr
            className="timeline-row"
          >
            <td>
              <a
                gatsby="true"
                href="/learn/relational-database/learn-sql-by-building-a-student-database-part-2/build-a-student-database-part-2"
              >
                Build a Student Database: Part 2
              </a>
            </td>
            <td />
            <td
              className="text-center"
            >
              <time
                dateTime="2016-09-28T20:31:56.730Z"
              >
                Dec 29, 2022
              </time>
            </td>
          </tr>
          <tr
            className="timeline-row"
          >
            <td>
              <a
                gatsby="true"
                href="/learn/relational-database/learn-sql-by-building-a-student-database-part-1/build-a-student-database-part-1"
              >
                Build a Student Database: Part 1
              </a>
            </td>
            <td />
            <td
              className="text-center"
            >
              <time
                dateTime="2016-09-28T20:31:56.730Z"
              >
                Dec 29, 2022
              </time>
            </td>
          </tr>
          <tr
            className="timeline-row"
          >
            <td>
              <a
                gatsby="true"
                href="/learn/relational-database/learn-bash-scripting-by-building-five-programs/build-five-programs"
              >
                Build Five Programs
              </a>
            </td>
            <td />
            <td
              className="text-center"
            >
              <time
                dateTime="2016-09-28T20:31:56.730Z"
              >
                Dec 29, 2022
              </time>
            </td>
          </tr>
          <tr
            className="timeline-row"
          >
            <td>
              <a
                gatsby="true"
                href="/learn/relational-database/build-a-celestial-bodies-database-project/build-a-celestial-bodies-database"
              >
                Build a Celestial Bodies Database
              </a>
            </td>
            <td />
            <td
              className="text-center"
            >
              <time
                dateTime="2016-09-28T20:31:56.730Z"
              >
                Dec 29, 2022
              </time>
            </td>
          </tr>
          <tr
            className="timeline-row"
          >
            <td>
              <a
                gatsby="true"
                href="/learn/relational-database/learn-relational-databases-by-building-a-mario-database/build-a-mario-database"
              >
                Build a Mario Database
              </a>
            </td>
            <td />
            <td
              className="text-center"
            >
              <time
                dateTime="2016-09-28T20:31:56.730Z"
              >
                Dec 29, 2022
              </time>
            </td>
          </tr>
          <tr
            className="timeline-row"
          >
            <td>
              <a
                gatsby="true"
                href="/learn/relational-database/learn-bash-by-building-a-boilerplate/build-a-boilerplate"
              >
                Build a Boilerplate
              </a>
            </td>
            <td />
            <td
              className="text-center"
            >
              <time
                dateTime="2016-09-28T20:31:56.730Z"
              >
                Dec 29, 2022
              </time>
            </td>
          </tr>
        </tbody>
      </table>
      <nav
        aria-label="aria.timeline-pagination-nav"
      >
        <ul
          aria-hidden="true"
          className="timeline-pagination_list"
        >
          <li
            className="timeline-pagination_list_item"
            style={
              {
                "visibility": "hidden",
              }
            }
          >
            <button
              aria-label="aria.first-page"
              disabled={true}
              onClick={[Function]}
            >
              &lt;&lt;
            </button>
          </li>
          <li
            className="timeline-pagination_list_item"
            style={
              {
                "visibility": "hidden",
              }
            }
          >
            <button
              aria-label="aria.previous-page"
              disabled={true}
              onClick={[Function]}
            >
              &lt;
            </button>
          </li>
          <li
            className="timeline-pagination_list_item"
          >
            profile.page-number
          </li>
          <li
            className="timeline-pagination_list_item"
            style={
              {
                "visibility": "unset",
              }
            }
          >
            <button
              aria-label="aria.next-page"
              disabled={false}
              onClick={[Function]}
            >
              &gt;
            </button>
          </li>
          <li
            className="timeline-pagination_list_item"
            style={
              {
                "visibility": "unset",
              }
            }
          >
            <button
              aria-label="aria.last-page"
              disabled={false}
              onClick={[Function]}
            >
              &gt;&gt;
            </button>
          </li>
        </ul>
      </nav>
    </section>
  </div>
</div>
`;
